### DISCUSSION TITLE
Error Loading Models from Pickle File with Custom Transformer in Sktime Pipeline

### DISCUSSION THREAD

**Discussion by RahulDas-dev** [alan-turing-institute/sktime#4920] [Q&A]
**Description**
I am currently facing an issue when trying to load models from a pickle file in Sktime. The pickle file contains a Sktime Pipeline, and within the Pipeline, there is a custom transformer that I created using the [extension template](https://github.com/sktime/sktime/blob/main/extension_templates/transformer.py)

**Error Message**
When attempting to load the models from the pickle file using `joblib.load()` , I encounter the following error:

```
AttributeError: Can't get attribute 'ColumnsGuard' on <module '__main__'>
```
However without the custom transformer it is behaving as expected.

**Custom Transformer**

The transformer *ColumnsGuard*  Verifies column names at predict time match the ones used when fitting

```
import warnings
from sktime.transformations.base import BaseTransformer


class ColumnsGuard(BaseTransformer):
    _tags = {
        "scitype:transform-input": "Series",
        "scitype:transform-output": "Series",
        "scitype:transform-labels": "None",
        "scitype:instancewise": True,  
        "capability:inverse_transform": True,  
        "univariate-only": False,  
        "X_inner_mtype": "pd.DataFrame",  
        "y_inner_mtype": "None",  
        "requires_y": False,  
        "enforce_index_type": None,  
        "fit_is_empty": False, 
        "X-y-must-have-same-index": True, 
        "transform-returns-same-time-index": True,
        "skip-inverse-transform": False,  
        "capability:unequal_length": True,
        "capability:unequal_length:removes": True,
        "handles-missing-data": False,  
        "capability:missing_values:removes": False,
        
    }

    def __init__(self):
        super(ColumnsGuard, self).__init__()


    def _fit(self, X, y=None):
        self._columns_name = X.columns.tolist()
        return self

    def _transform(self, X, y=None):
        if self._columns_name == X.columns.tolist():
            X_transformed = X.copy(deep=True)
        else:
            missing_col = set(self._columns_name) - set(X.columns.tolist())
            extra_col = set(X.columns.tolist()) - set(self._columns_name)  
            if missing_col:
                raise ValueError('Missing columns: {missing_col}') 
            elif extra_col:
                warnings.warn('Got extra columns: {extra_col}, ignoring')
                X_transformed = X[self.self._columns_name].copy(deep=True)
        return X_transformed  

    def _inverse_transform(self, X, y=None):
        X_inv_transformed = X.copy(deep=True)
        return X_inv_transformed     

    @classmethod
    def get_test_params(cls, parameter_set="default"):
        params = {}
        return params
    
    
from sktime.utils.estimator_checks import check_estimator

results = check_estimator(ColumnsGuard, raise_exceptions=True)    
```
`All tests PASSED!`

**The Forecaster Pipeline**

*Step 1* Model Building
```
from sktime.datasets import load_macroeconomic
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.compose import ForecastingPipeline

from sktime.forecasting.model_selection import temporal_train_test_split

from sklearn.ensemble import RandomForestRegressor

forecaster_pipe = ForecastingPipeline(
    steps=[
        ('column_Gaurd', ColumnsGuard()),
        ( "forecaster",  make_reduction(
                            estimator=RandomForestRegressor(random_state=42),
                            scitype="tabular-regressor",
                            window_length=10,
                            strategy="recursive",
                            pooling="global",
                        ),
        )
    ]           
)


data = load_macroeconomic()
train_xy, test_xy = temporal_train_test_split(data, test_size=5)

train_y = train_xy.pop('unemp')
#train_xy.head()

test_y = test_xy.pop('unemp')

forecaster_pipe.fit(train_y,train_xy, fh=[1,2,3,4,5])

y_pred = forecaster_pipe.predict(fh=[1,2,3,4,5], X=test_xy)

print(y_pred )

import joblib

joblib.dump(forecaster_pipe, "../../results/RandomForestRegressor.pkl")
```
*Step -2* Loading the trained model [ This was executed in separate notebook ]

```
import joblib

trained_model = joblib.load("../../results/RandomForestRegressor.pkl")
```

**The Error Stack**

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[1], line 3
      1 import joblib
----> 3 trained_model = joblib.load("../../results/RandomForestRegressor.pkl")

File ~\OneDrive\Desktop\codespace\Python\timeseries\Time-Series-Examples\.venv\lib\site-packages\joblib\numpy_pickle.py:658, in load(filename, mmap_mode)
    652             if isinstance(fobj, str):
    653                 # if the returned file object is a string, this means we
    654                 # try to load a pickle file generated with an version of
    655                 # Joblib so we load it with joblib compatibility function.
    656                 return load_compatibility(fobj)
--> 658             obj = _unpickle(fobj, filename, mmap_mode)
    659 return obj

File ~\OneDrive\Desktop\codespace\Python\timeseries\Time-Series-Examples\.venv\lib\site-packages\joblib\numpy_pickle.py:577, in _unpickle(fobj, filename, mmap_mode)
    575 obj = None
    576 try:
--> 577     obj = unpickler.load()
    578     if unpickler.compat_mode:
    579         warnings.warn("The file '%s' has been generated with a "
    580                       "joblib version less than 0.10. "
    581                       "Please regenerate this pickle file."
    582                       % filename,
    583                       DeprecationWarning, stacklevel=3)

File ~\AppData\Local\Programs\Python\Python39\lib\pickle.py:1212, in _Unpickler.load(self)
   1210             raise EOFError
   1211         assert isinstance(key, bytes_types)
-> 1212         dispatch[key[0]](self)
   1213 except _Stop as stopinst:
   1214     return stopinst.value

File ~\AppData\Local\Programs\Python\Python39\lib\pickle.py:1537, in _Unpickler.load_stack_global(self)
   1535 if type(name) is not str or type(module) is not str:
   1536     raise UnpicklingError("STACK_GLOBAL requires str")
-> 1537 self.append(self.find_class(module, name))

File ~\AppData\Local\Programs\Python\Python39\lib\pickle.py:1581, in _Unpickler.find_class(self, module, name)
   1579 __import__(module, level=0)
   1580 if self.proto >= 4:
-> 1581     return _getattribute(sys.modules[module], name)[0]
   1582 else:
   1583     return getattr(sys.modules[module], name)

File ~\AppData\Local\Programs\Python\Python39\lib\pickle.py:331, in _getattribute(obj, name)
    329         obj = getattr(obj, subpath)
    330     except AttributeError:
--> 331         raise AttributeError("Can't get attribute {!r} on {!r}"
    332                              .format(name, obj)) from None
    333 return obj, parent

AttributeError: Can't get attribute 'ColumnsGuard' on <module '__main__'>

```

**Additional Information**

- Sktime version: 0.20.0
- Python version: 3.9.7

**Note**
The step 2 is executed in separate notebook  .

I have already checked the Sktime documentation and GitHub issues but couldn't find a solution that addresses this specific problem.

Any assistance or insights on resolving this issue would be greatly appreciated. Thank you!

**Comment by fkiraly** [comment]
Hm, interesting - thanks for reporting, @RahulDas-dev - I think this is a bug.

We may not have thought through the interaction between custom transformers and the serialization framework.

The problem seems to be that the custom estimator isn't saved as a full specification, but only as a reference - and the reference doesn't seem to be present in the environment.

Here's an interesting question now: should we save custom estimators with their full code?
Or should we assume presence in some registered 3rd party module?

FYI @achieveordie

**Reply by achieveordie** [reply]
Allowing to save custom estimators with their complete implementation sounds like a risky deal IMO. It would make the most sense to assume the custom estimator is already present in the current heap, missing that should raise the above-mentioned error.

**Comment by RahulDas-dev** [comment]
Thank you for the prompt response. I was also contemplating the issue of the missing reference in the environment. Currently, I am `importing the custom transformer` wherever I load it from pickle .

Both the `TransformedTargetForecaster` and `ForecastingPipeline` work seamlessly in the serialization framework. I believe we should aim for the same for the custom transformer.

I would like to raise a pull request (PR) if you could provide me with some guidance on same.

**Reply by fkiraly** [reply]
Thanks!

I'm not quite sure how to solve this best - but the logic currently sits in `BaseObject`'s `load_from_file` and `load_from_serial` for loading, and `save` for saving.

I can think of two ways to deal with this:
* storing the component definition with the component itself, i.e., the source code.
* ensuring the loader has a reference to the 3rd party class, e.g., by a registration utility for 3rd party classes.

Option 1 could be a security risk though, as this can now be any python code that is executed - so I would tend towards option 2.

Any additional ideas, @achieveordie?
Are you serializing your models too, @yarnabrina?

**Reply by yarnabrina** [reply]
> Are you serializing your models too

I am, but I do not use custom transformers with 3rd party (is `sklearn` 3rd party, as it's a dependency?) packages yet.

My current work involves few `sklearn` transformers through `TabularToSeriesAdaptor` and using it along with `FeatureUnion` and `ForecasterPipeline`. Then I dump (and load) using `dill`, so far haven't faced problems in same or different sessions.

**Reply by fkiraly** [reply]
> is sklearn 3rd party, as it's a dependency

`sklearn` is a core depedecy, so its estimators are always in the import scope.

> Then I dump (and load) using dill

Would it make sense to allow users to select the serialization ackend in the save/load methods?

**Reply by yarnabrina** [reply]
I'm not sure if this choice will be very meaningful or popular with users, but I do not have anything against it.

**Reply by achieveordie** [reply]
> Would it make sense to allow users to select the serialization backend in the save/load methods?

I like the idea!
